src/protocol/mod.rs

```diff
@@ -1,65 +1,145 @@
+/// Step "a" would be executed with a context identifier of "protocol/a", which it
+///  would `refine()` into "protocol/a/x" and "protocol/a/y" to produce a final set
+/// of identifiers: ".../a/x", ".../a/y", ".../b", and ".../c".
 ///
-/// See `IPAProtocolStep` for a canonical implementation of this trait. Every time we switch to
-/// use a new circuit, there will be an additional struct/enum that implements `Step`, but eventually
-/// it should converge to a single implementation.
-pub trait Step: Copy + Clone + Debug + Eq + Hash + Send + 'static {}
+/// Note that the implementation of this context might change to use a different
+/// (possible more efficient) representation.  It is probably not particularly efficient
+/// to be cloning this object all over the place.  Of course, a string is pretty useful
+/// from a debugging perspective.
+#[derive(Clone, Debug)]
+pub struct UniqueStepId {
+    id: String,
+    /// This tracks the different values that have been provided to `refine()`.
+    #[cfg(debug_assertions)]
+    used: Arc<Mutex<HashSet<String>>>,
+}
 
-/// Set of steps that define the IPA protocol.
-#[derive(Copy, Clone, PartialEq, Eq, Hash)]
-pub enum IPAProtocolStep {
-    /// Convert from XOR shares to Replicated shares
-    ConvertShares,
-    /// Sort shares by the match key
-    Sort(SortStep),
+unsafe impl Send for UniqueStepId {}
+impl Hash for UniqueStepId {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(self.id.as_bytes());
+    }
 }
+impl PartialEq for UniqueStepId {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+impl Eq for UniqueStepId {}
 
-impl Debug for IPAProtocolStep {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "IPA/")?;
-        match self {
-            IPAProtocolStep::ConvertShares => {
-                write!(f, "ConvertShares")
-            }
-            IPAProtocolStep::Sort(sort_step) => {
-                write!(f, "Sort/{:?}", sort_step)
-            }
+impl UniqueStepId {
+    /// Narrow the scope of the step identifier.
+    /// # Panics
+    /// In a debug build, this checks that the same refine call isn't run twice and that the string
+    /// value of the step doesn't include '/' (which would lead to a bad outcome).
+    #[must_use]
+    pub fn narrow<S: Step>(&self, step: &S) -> Self {
+        #[cfg(debug_assertions)]
+        {
+            let s = String::from(step.as_ref());
+            assert!(!s.contains('/'), "The string for a step cannot contain '/'");
+            assert!(
+                self.used.lock().unwrap().insert(s),
+                "Refined '{}' with step '{}' twice",
+                self.id,
+                step.as_ref(),
+            );
+        }
+
+        Self {
+            id: self.id.clone() + "/" + step.as_ref(),
+            #[cfg(debug_assertions)]
+            used: Arc::new(Mutex::new(HashSet::new())),
```

Comment:
You probably wanted to clone it from `self`

Reply:
No, this was deliberate.  Once this is refined down to the next level, you want a clean slate.

Reply:
meaning you might want to reuse steps at different levels?